msal/src/Microsoft.Identity.Client/Features/DeviceCode/DeviceCodeRequest.cs

```diff
@@ -0,0 +1,133 @@
+﻿//------------------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.Internal.Requests;
+using Microsoft.Identity.Core.Helpers;
+using Microsoft.Identity.Core.OAuth2;
+
+namespace Microsoft.Identity.Client.Features.DeviceCode
+{
+    internal class DeviceCodeRequest : RequestBase
+    {
+        private readonly Func<DeviceCodeResult, Task> _deviceCodeResultCallback;
+        private DeviceCodeResult _deviceCodeResult;
+
+        public DeviceCodeRequest(
+            AuthenticationRequestParameters authenticationRequestParameters, 
+            Func<DeviceCodeResult, Task> deviceCodeResultCallback)
+            : base(authenticationRequestParameters)
+        {
+            _deviceCodeResultCallback = deviceCodeResultCallback;
+
+            LoadFromCache = false;  // no cache lookup for token
+            SupportADFS = false;
+            StoreToCache = true;
+        }
+
+        internal override async Task PreTokenRequestAsync(CancellationToken cancellationToken)

```

Comment:
should IsCancellationRequested be queried after each of the async calls made here?

Reply:
Eventually, yes.  Given we talked about not exposing cancellation to anyone else, the main place you'll want to cancel devicecode is in the polling state which happens within only one of these methods.  so I think we should do the "iscancellationrequested" check once we've more fully embraced cancellation in the main methods.